src/Voron/Data/Tables/TableValueCompressor.cs

```diff
@@ -229,6 +229,18 @@ private void MaybeTrainCompressionDictionary(Table table, FixedSizeTree etagsTre
                         }
 
                         tx.LowLevelTransaction.OnDispose += RecreateRecoveryDictionaries;
+                        tx.LowLevelTransaction.OnDispose += state =>
+                        {
+                            if (state is not LowLevelTransaction llt || llt.Committed)
+                                return;
+                            // ***************************************
+                            // RavenDB-17758  - This is *important* - if we aren't committing the transaction, we *have*
+                            // to remove the in memory cache for the dictionary, since it wasn't written to disk. If we
+                            // keep it, we can compress data with a dictionary that we rolled back, and end up with data
+                            // corruption!!!
+                            // ****************************************
+                            tx.LowLevelTransaction.Environment.CompressionDictionariesHolder.Remove(newId);
```

Comment:
In my humble opinion, it will be safer to add the new dictionary to the global memory state only on success
instead of removing it on fail.
I tried to find a way to fail this implementation but maybe my test is too aggressive
but even if the current implementation works future leak can cause it again.

Let's say we have three commands on the transaction merger:
1. Will fail on out of memory in the `LowLevelTransaction.OnDispose`.
2. Create a new compression dictionary.
3. Fail for any reason (and abort the merged transaction).

```
[Fact]
public async Task TestCase()
{
    // var server = GetNewServer(new ServerCreationOptions {RunInMemory = false,});
    using var store = GetDocumentStore(new Options { RunInMemory = false});
    
    var record = store.Maintenance.Server.Send(new GetDatabaseRecordOperation(store.Database));
    record.DocumentsCompression = new DocumentsCompressionConfiguration(compressRevisions: true, "TestObjs");
    store.Maintenance.Server.Send(new UpdateDatabaseOperation(record, record.Etag));
    
    using (var session = store.OpenAsyncSession())
    {
        for (int i = 0; i < 1024; i++)
        {
            await session.StoreAsync(new TestObj(), $"testObjs/{i}");
        }
        await session.SaveChangesAsync();
    }

    var database = await GetDatabase(store.Database);
    using (database.DocumentsStorage.ContextPool.AllocateOperationContext(out JsonOperationContext context))
    {
        var data = new DynamicJsonValue
        {
            [Constants.Documents.Metadata.Key] = new DynamicJsonValue {[Constants.Documents.Metadata.Collection] = "TestObjs"}
        };
        var mergedTransactionCommand = new BatchHandler.MergedBatchCommand(database)
        {
            ParsedCommands = new ArraySegment<BatchRequestParser.CommandData>(new []
            {
                new BatchRequestParser.CommandData
                {
                    Type = CommandType.PUT,
                    Id = "testObjs/00",
                    Document = context.ReadObject(data, "json")
                },
                new BatchRequestParser.CommandData
                {
                    Type = CommandType.PUT,
                    Id = "testObjs/000",
                    Document = context.ReadObject(data, "json")
                }
            })
        };
        await Task.WhenAll(
            database.TxMerger.Enqueue(new OutOfMemoryOnDisposeCommand()).ContinueWith(_ => string.Empty),
            database.TxMerger.Enqueue(mergedTransactionCommand).ContinueWith(_ => string.Empty),
            database.TxMerger.Enqueue(new FailedCommand()).ContinueWith(_ => string.Empty)
        );
    }
}

private class TestObj
{
    public string Id { get; set; }
    public string Prop { get; set; }
}

private class TestException : Exception
{

}

private class FailedCommand : TransactionOperationsMerger.MergedTransactionCommand
{
    protected override long ExecuteCmd(DocumentsOperationContext context) => throw new TestException();

    public override TransactionOperationsMerger.IReplayableCommandDto<TransactionOperationsMerger.MergedTransactionCommand> ToDto(JsonOperationContext context) =>
        throw new NotImplementedException();
}
private class OutOfMemoryOnDisposeCommand : TransactionOperationsMerger.MergedTransactionCommand
{
    protected override long ExecuteCmd(DocumentsOperationContext context)
    {
        context.Transaction.InnerTransaction.LowLevelTransaction.OnDispose += _ => throw new OutOfMemoryException();
        return 1;
    }

    public override TransactionOperationsMerger.IReplayableCommandDto<TransactionOperationsMerger.MergedTransactionCommand> ToDto(JsonOperationContext context) =>
        throw new NotImplementedException();
}

```

Reply:
Not adding the compression dictionary to the cache is actually quite hard, because we don't have a way of knowing what transaction it was added.

Also, for your code, aside from the move to the `Rollback`, I would expect this to be a fatal error that should force us to handle this properly.